gateway/src/apicast/policy.lua

```diff
@@ -30,6 +30,29 @@ local function __eq(policy, other)
     return policy._NAME == other._NAME and policy._VERSION == other._VERSION
 end
 
+local inspect = require('inspect')
+
+local function __gc(proxy)
+    local policy = getmetatable(proxy).__index -- proxy always points to the policy
+    local mt = getmetatable(policy) or empty-- in case someone changed the policy metatable
+
+    local gc = mt.__gc or noop
+
+    return gc(policy)
+end
+
+local function metatable(policy)
+    local proxy = newproxy(true)
```

Comment:
To be honest, I learnt about `newproxy` while reviewing this PR.

If I understood it correctly we need to do this to include `__gc` in the table. It looks like this is not needed in Lua 5.2+ and will not be needed in future versions of LuaJIT: https://github.com/LuaJIT/LuaJIT/issues/47

I think this merits a comment.

Reply:
It definitely needs plenty of comments. 

I seriously doubt it will land in LuaJIT because it depends on https://github.com/LuaJIT/LuaJIT/issues/38. Even though LuaJIT/LuaJIT#47 is closed it is not implemented.

Another approach is to use void pointer from ffi and use ffi.gc like:

```lua
local ffi = require('ffi')
local p = ffi.new('void *')
 ffi.gc(p, function() print('__gc') end)
collectgarbage()
p = nil
collectgarbage()
```